A **Genetic Algorithm** can be used to solve real-world business problems like **Merchandise Optimization**. In large stores like Target or Walmart, placing products on shelves involves complex planning and strategy. With hundreds of aisles and countless product combinations, GAs can help find the best arrangement to maximize efficiency and sales.

Now imagine instead of filling up a knapsack, you have to properly allocate millions of dollars worth of resources for a particular project, optimize the inventory for different stages of an extensive supply chain.

As mentioned previously, a **Genetic Algorithm is inspired by natural selection**. Natural Selection involves several individuals (a population) interacting and living within a particular environment. Over time, individuals with traits better suited for their specific environment will tend to last from generation to generation, while individuals with less suitable characteristics will die out.

### The Scenario

You are a data scientist for a retail company. The company just opened a micro store and has 50 products they would like to stock. Space is limited, so given that each product occupies a certain amount of space in addition to an estimated profit, you are tasked with finding the most profitable set of products to stock, given the space constraints.

### Solution — Create a Genetic Algorithm - Stepwise

Knowing the countless solutions to this problem, you elect to utilize a genetic algorithm. Below is a list of the available products that can be stocked. Take note that our micro store only has 100 units of shelf space.

### Code 1: Retail Store Items


This part of the code defines a dictionary called retail_store_items. It lists various items that a retail store might sell, along with two pieces of information for each item:

1. Shelf Space: How much space the item needs on the shelf.

2. Expected Profit: How much profit the store expects to make from selling that item.

          retail_store_items = {
              "Toothpaste": {"shelf_space": 1, "expected_profit": 2.50},
              "Shampoo": {"shelf_space": 3, "expected_profit": 3.75},
              ...  }


In [26]:
retail_store_items = {
    "Toothpaste": {"shelf_space": 1, "expected_profit": 2.50},
    "Shampoo": {"shelf_space": 3, "expected_profit": 3.75},
    "Conditioner": {"shelf_space": 3, "expected_profit": 3.50},
    "Body Wash": {"shelf_space": 3, "expected_profit": 3.00},
    "Deodorant": {"shelf_space": 1, "expected_profit": 2.20},
    "Lotion": {"shelf_space": 3, "expected_profit": 3.00},
    "Shaving Cream": {"shelf_space": 2, "expected_profit": 2.50},
    "Razor Blades": {"shelf_space": 1, "expected_profit": 4.50},
    "Mouthwash": {"shelf_space": 2, "expected_profit": 2.00},
    "Hand Soap": {"shelf_space": 1, "expected_profit": 1.50},
    "Facial Tissue": {"shelf_space": 2, "expected_profit": 1.20},
    "Paper Towels": {"shelf_space": 5, "expected_profit": 3.00},
    "Toilet Paper": {"shelf_space": 5, "expected_profit": 2.80},
    "Dish Soap": {"shelf_space": 3, "expected_profit": 2.75},
    "Laundry Detergent": {"shelf_space": 6, "expected_profit": 4.50},
    "Fabric Softener": {"shelf_space": 3, "expected_profit": 3.50},
    "Bleach": {"shelf_space": 4, "expected_profit": 2.00},
    "All-Purpose Cleaner": {"shelf_space": 3, "expected_profit": 2.50},
    "Glass Cleaner": {"shelf_space": 2, "expected_profit": 2.00},
    "Furniture Polish": {"shelf_space": 2, "expected_profit": 2.50},
    "Trash Bags": {"shelf_space": 4, "expected_profit": 3.00},
    "Light Bulbs": {"shelf_space": 2, "expected_profit": 1.50},
    "Batteries": {"shelf_space": 1, "expected_profit": 2.00},
    "Extension Cords": {"shelf_space": 3, "expected_profit": 4.00},
    "Pet Food": {"shelf_space": 6, "expected_profit": 3.50},
    "Cat Litter": {"shelf_space": 5, "expected_profit": 4.00},
    "Dog Leash": {"shelf_space": 2, "expected_profit": 3.00},
    "Pet Shampoo": {"shelf_space": 2, "expected_profit": 2.50},
    "Pet Toys": {"shelf_space": 2, "expected_profit": 1.50},
    "Shower Curtain": {"shelf_space": 4, "expected_profit": 3.00},
    "Bath Towels": {"shelf_space": 4, "expected_profit": 4.00},
    "Bed Sheets": {"shelf_space": 6, "expected_profit": 8.00},
    "Pillows": {"shelf_space": 5, "expected_profit": 5.00},
    "Blankets": {"shelf_space": 6, "expected_profit": 6.00},
    "Curtains": {"shelf_space": 5, "expected_profit": 7.00},
    "Wall Clocks": {"shelf_space": 3, "expected_profit": 4.50},
    "Picture Frames": {"shelf_space": 2, "expected_profit": 2.00},
    "Candles": {"shelf_space": 2, "expected_profit": 2.50},
    "Vases": {"shelf_space": 3, "expected_profit": 3.00},
    "Decorative Pillows": {"shelf_space": 4, "expected_profit": 3.50},
    "Alarm Clocks": {"shelf_space": 2, "expected_profit": 2.00},
    "Phone Chargers": {"shelf_space": 1, "expected_profit": 3.00},
    "USB Cables": {"shelf_space": 1, "expected_profit": 1.50},
    "Headphones": {"shelf_space": 2, "expected_profit": 5.00},
    "Laptop Sleeves": {"shelf_space": 2, "expected_profit": 4.00},
    "Notebooks": {"shelf_space": 2, "expected_profit": 1.50},
    "Pens": {"shelf_space": 1, "expected_profit": 1.00},
    "Sticky Notes": {"shelf_space": 1, "expected_profit": 1.00},
    "Folders": {"shelf_space": 2, "expected_profit": 1.50},
    "Scissors": {"shelf_space": 1, "expected_profit": 2.00}
}


**Object Oriented Programming**


To execute this solution, we will build two classes in Python: one that represents a combination of the available products we can choose from and one for our genetic algorithm to help us converge on the optimal combination of products to stock in the store.

In [27]:
import random
from tqdm import tqdm

### Individual Class

As I just mentioned, our individual class will represent a combination of products from the retail_store_items dictionary. This class will have the following attributes:

1. merch: a nested dictionary of the available items we can choose from.
max_shelf_space: the maximum amount of shelf space our selection of items can occupy.

2. chromosome: a list of 1s and 0s that will represent whether or not a product from merch is included. Each chromosome will have the same length of merch. The 1s and 0s will be generated at random.

3. occupied_space: the amount of space occupied by the selected subset of merch. This will be calculated using a fitness function.

4. total_expected_profit: the total amount of expected profit by the selected subset of merch. This will be calculated using a fitness function.

5. generation: This will be a placeholder for an individual's generation once this class is incorporated into the genetic algorithm.

In [28]:
class Individual():
    def __init__(self, merch, max_shelf_space):
        self.merch = merch
        self.max_shelf_space = max_shelf_space
        self.chromosome = [random.choice([0, 1]) for _ in range(len(merch))]
        self.occupied_space = 0
        self.total_expected_profit = 0
        self.generation = 0
        self.fitness()  # Ensure fitness is called during initialization


**The Fitness method**


Our fitness function will calculate two attributes for us: the occupied space and the total expected profit.

To execute this, we will loop through the chromosome, use each 1 or 0 as a multiplier against each item's occupied space and expected profit, and add them to the attributes we mentioned.

Note that total_expected_profit will automatically be 0 if our subset of products exceeds max_shelf_space.

Remember, this will be the heart of our genetic algorithm; it is the driving force that will determine which combinations of products will permeate through generations to converge on the best solution.

In [29]:
    def fitness(self):
        self.occupied_space = 0
        self.total_expected_profit = 0
        for (item, values), multiplier in zip(self.merch.items(), self.chromosome):
            self.occupied_space += values['shelf_space'] * multiplier
            self.total_expected_profit += values['expected_profit'] * multiplier
            if self.occupied_space > self.max_shelf_space:
                self.total_expected_profit = 0
                break


The Crossover method


Next, we will build our crossover function. This function will take two separate instances of an individual class, randomly generate two cutoff points for each instance's chromosome, and use it to create four new individual classes, essentially the children of these two individuals.

Earlier, we discussed how the crossover would make two children. I'm building this function to ensure we have four because when we develop our genetic algorithm, I want a more comprehensive selection of child instances to permeate the next generation.

Also, note how our generation attribute for each child would increase by 1 when this function is called.

In this code, a crossover operation is performed between two "individuals" (you and the "partner"). This is a key part of the genetic algorithm where two parents combine parts of their data (chromosomes) to create new "children." Here's what happens step by step:

1. Two cutoff points are chosen randomly in the chromosome (using cutoff_1 and cutoff_2), which determines where the chromosomes will be split.

2. Four new children (child1, child2, child3, child4) are created:

    2.1 Child 1 and Child 2 are created using the first cutoff point (cutoff_1):

        2.1.1 Child 1 gets the start of the partner's chromosome and the end of the self's chromosome.

        2.1.2 Child 2 gets the start of the self's chromosome and the end of the partner's chromosome.
        
    2.2 Child 3 and Child 4 are created similarly, but using the second cutoff point (cutoff_2).

3. After forming the children, each child's fitness is calculated, which evaluates how good their solution is.

The generation of each child is incremented by 1, meaning they belong to the next generation compared to the parent.

Finally, the method returns the four children to continue the genetic algorithm process.

This process simulates reproduction, where different combinations of chromosomes create variation in the new population.

In [30]:
    def crossover(self, partner):
        cutoff_1 = random.randint(1, len(self.chromosome) - 1)
        cutoff_2 = random.randint(1, len(self.chromosome) - 1)

        child1 = Individual(self.merch, self.max_shelf_space)
        child2 = Individual(self.merch, self.max_shelf_space)
        child3 = Individual(self.merch, self.max_shelf_space)
        child4 = Individual(self.merch, self.max_shelf_space)

        child1.chromosome = partner.chromosome[:cutoff_1] + self.chromosome[cutoff_1:]
        child2.chromosome = self.chromosome[:cutoff_1] + partner.chromosome[cutoff_1:]
        child3.chromosome = partner.chromosome[:cutoff_2] + self.chromosome[cutoff_2:]
        child4.chromosome = self.chromosome[:cutoff_2] + partner.chromosome[cutoff_2:]

        child1.fitness()
        child2.fitness()
        child3.fitness()
        child4.fitness()

        child1.generation = self.generation + 1
        child2.generation = self.generation + 1
        child3.generation = self.generation + 1
        child4.generation = self.generation + 1

        return child1, child2, child3, child4


The Mutation method


As our algorithm progresses, there is a chance it won't converge appropriately because it didn't explore other combinations besides the ones that have been reinforced by our fitness and crossover function between generations.

 This is very similar to what happens when a neural network's learning rate is too high. This is where the mutation function comes in. Given a probability, our mutation function will randomly change one of an individual's chromosomes.

It's as simple as that. When we code our genetic algorithm class, we will call the mutation function right after the crossover so that each generation's children can have a chance at forming a new attribute.

In this code, a mutation operation is being performed on the individual's chromosome with a certain probability. Here's a breakdown of what happens:

1. Mutation probability: The random.random() function generates a random number between 0 and 1. If this number is less than the rate, a mutation occurs. The rate is the probability of mutation, controlling how often mutations happen.

2. Selecting a position: A random position in the chromosome is selected using random.randint(). This position (random_position) is where the mutation will take place.

3. Flipping the gene: At the chosen position, the value of the gene (bit) in the chromosome is flipped:

    If the gene is 1, it becomes 0.
    If the gene is 0, it becomes 1. This simulates a mutation, introducing a small random change to the individual.

4. Recalculate fitness: After the mutation, the individual's fitness is recalculated to see how the mutation affects the overall performance of this solution.

In essence, this code introduces a small random change (mutation) to the individual's chromosome, which is key to maintaining genetic diversity in a population.

In [31]:
    def mutation(self, rate):
        if random.random() < rate:
            random_position = random.randint(0, len(self.chromosome) - 1)
            self.chromosome[random_position] = 1 - self.chromosome[random_position]
            self.fitness()


View Stock Function

I decided to add this optional function out of curiosity. It lets us quickly see the subset of products for an individual instance.


In this view_shelf method, the code is iterating over the items in the self.merch dictionary and printing certain items based on values in the self.chromosome list. Here’s what happens step by step:

1. Iterating through items and chromosomes:

    The method uses a for loop with zip() to loop over both self.merch.items() and self.chromosome at the same time.

    self.merch.items() refers to the dictionary items where item is the key (e.g., the name of a product) and values is the associated value (e.g., details of the product).

    self.chromosome is a list of binary values (0s and 1s) representing whether an item should be included on the shelf.

2. Checking the multiplier (chromosome):

    For each item in self.merch, the corresponding value in self.chromosome (called multiplier) is checked.

    If the multiplier is 1, the item is printed, meaning it is "selected" to be shown on the shelf.

3. Printing selected items:

    Only items with a corresponding multiplier of 1 are printed, representing which items are chosen to be on the shelf (likely as part of the genetic algorithm's solution).


In summary, this method prints the items from self.merch where the corresponding value in self.chromosome is 1, effectively showing a "shelf" of selected items based on the genetic algorithm's chromosome.

In [32]:
    def view_shelf(self):
        for (item, values), multiplier in zip(self.merch.items(), self.chromosome):
            if multiplier == 1:
                print(item)


### Genetic Algorithm Class


The stage is set to build our Genetic Algorithm class. Ultimately, this class will run the natural selection simulation and store the best-performing individuals as they are identified. This class will have the following attributes:

    population_size: the size of each population within each iteration of the algorithm.

    num_of_generations: the total number of iterations to run.

    mutation_rate: the mutation rate to be applied to each child when the crossover function is called in the individual class.

    max_shelf_space: identical to the same attribute in the individual class.

    items: the nested dictionary of available items to choose from.

    best_expected_profit: the placeholder for the best total_expected_profit recorded from all iterations.

    best_Individuals: a list where each subsequent Individual class is added to when it is revealed it beats the latest best_expected_profit.

    population: a list where each generation's population will be applied to.

In [33]:
class GeneticAlgorithm():
    def __init__(self, population_size, num_of_generations, mutation_rate, max_shelf_space, items):
        self.population_size = population_size
        self.num_of_generations = num_of_generations
        self.mutation_rate = mutation_rate
        self.max_shelf_space = max_shelf_space
        self.items = items
        self.best_expected_profit = 0
        self.best_individuals = []
        self.population = []


The run_simulation method


Looking at the code below, this method may look complicated, but it becomes straightforward once you break it down. It starts by creating an initial population of Individual objects. From this population, we will check to see if each object's total_expected_profit is higher than the best_expected_profit attribute.


After the initial population is created, we will shuffle it randomly and apply the crossover function to every group of two. Shuffling at this stage and subsequent populations ensures we exploit as many combinations as possible. For each population, we will take every two individual objects in the population list and run the crossover method on them. Four children are created from this method, the mutation method is applied, and finally, we append to a list that we sort by the total_expected_profit. Only the top two children make it to the subsequent population, and check to see if they have a higher total_expected_profit than best_expected_profit. Ensuring the top two children make it to the next generation is key to ensuring we are continuing to converge on an even better solution.


This run_simulation method is running a genetic algorithm to find the best solution for maximizing profit in a merchandising problem. Here’s what happens step by step:

1. Initial Population:

1.1 A population of individuals is created.

1.2 For each individual:

  1.2.1 An Individual is initialized with merchandise (self.items) and a maximum shelf space (self.max_shelf_space).

  1.2.2 The individual's expected profit is calculated.

  1.2.3. The best individual(s) are tracked:

      If the individual's profit is greater than the current best profit, it becomes the new best.

      Otherwise, if the profit is equal to or above the lowest profit in the best group, they are added to the best group.

2. Subsequent Generations:

2.1 The simulation evolves over a set number of generations (self.num_of_generations), where the population improves over time.

2.2 The population is shuffled randomly at the start of each generation.

3. Crossover & Mutation:

3.1 For each pair of individuals (parents) in the shuffled population:

      3.1.1 A crossover is performed between two parents to create four children.

      3.1.2 Each child undergoes a mutation, where random changes are introduced based on the mutation rate (self.mutation_rate).

      3.1.3 The four children are sorted by their total expected profit.

      3.1.4 The top 2 children are selected as the best offspring and are added to the next generation.

      3.1.5 The best individual(s) are updated if any of the top children exceed the current best expected profit.

4. New Population:

4.1 After all pairs have produced offspring, the new population consists of the best 2 children from each pair.

4.2 The population is updated for the next generation.


**Purpose:**
The goal of this simulation is to maximize the total expected profit by evolving a population of potential solutions (individuals) through crossover, mutation, and selection. The best individuals are tracked and updated over multiple generations, with the genetic algorithm improving the population over time.

In [34]:
    def run_simulation(self):
        #### Initial Population ####
        for i in range(self.population_size):
            individual = Individual(merch=self.items, max_shelf_space=self.max_shelf_space)
            self.population.append(individual)
            if individual.total_expected_profit > self.best_expected_profit:
                self.best_expected_profit = individual.total_expected_profit
                self.best_individuals = [individual]
            else:
                self.best_individuals = [ind for ind in self.best_individuals if ind.total_expected_profit >= individual.total_expected_profit]
                self.best_individuals.append(individual)

        #### Subsequent Generations ####
        for g in tqdm(range(self.num_of_generations)):
            random.shuffle(self.population)
            new_population = []

            #### Crossover between each pair ####
            for pair in range(0, len(self.population), 2):
                parent_1 = self.population[pair]
                parent_2 = self.population[pair + 1]
                child_1, child_2, child_3, child_4 = parent_1.crossover(partner=parent_2)
                child_1.mutation(rate=self.mutation_rate)
                child_2.mutation(rate=self.mutation_rate)
                child_3.mutation(rate=self.mutation_rate)
                child_4.mutation(rate=self.mutation_rate)
                all_children = [child_1, child_2, child_3, child_4]
                all_children.sort(key=lambda x: x.total_expected_profit, reverse=True)
                top_children = all_children[:2]
                if top_children[0].total_expected_profit > self.best_expected_profit:
                    self.best_expected_profit = top_children[0].total_expected_profit
                    self.best_individuals = [top_children[0]]
                elif top_children[1].total_expected_profit > self.best_expected_profit:
                    self.best_expected_profit = top_children[1].total_expected_profit
                    self.best_individuals = [top_children[1]]
                new_population.append(top_children[0])
                new_population.append(top_children[1])

            #### Assign new population ####
            self.population = new_population


Running the Genetic Algorithm


We can now run our Genetic Algorithm! We should find a solution that makes the most of the available space and maximizes our expected profit. However, how will we know if our solution is a good one? The best solution I came up with is comparing our genetic algorithm to a baseline. To do this, we will create another class called RandomAlgorithm that generates Individual classes at random and checks to see which one has the best total_expected_profit. The code for this is below for reference.

In [35]:
class RandomAlgorithm():
    def __init__(self, iterations, max_shelf_space, items):
        self.iterations = iterations
        self.max_shelf_space = max_shelf_space
        self.items = items
        self.best_individuals = []
        self.best_expected_profit = 0

    def fitness(self):
        pass  # Implement fitness function logic here

    def run_simulation(self):
        #### Initial Population ####
        for i in tqdm(range(self.iterations)):
            trial = Individual(merch=self.items, max_shelf_space=self.max_shelf_space)
            trial.fitness()
            if trial.total_expected_profit > self.best_expected_profit:
                self.best_expected_profit = trial.total_expected_profit
                self.best_individuals = [trial]


## Combine code

This code simulates how a retail store can decide which items to place on a shelf to maximize profit, while also considering the space available. It uses a genetic algorithm and a random algorithm to find the best combination of items to place on the shelf.

1. Retail Store Items: This dictionary contains items like toothpaste and shampoo, with their required shelf space and expected profit.

2. Individual Class: Each "individual" represents a possible combination of items on the shelf. The chromosome (a list of 0s and 1s) shows whether an item is on the shelf (1) or not (0). The fitness() function checks if the individual is profitable without exceeding the shelf space.

3. Crossover: This method creates new individuals by mixing two parents' chromosomes (two existing combinations of items). Two children are produced, and their profitability is calculated.

4. Mutation: Occasionally, one item in the individual’s chromosome might flip from 0 to 1 (or vice versa), simulating small random changes.

5. GeneticAlgorithm Class: This runs the main simulation. It starts with a random population of individuals (combinations of items) and goes through generations, where parents are crossed over, mutated, and the best combinations survive to the next generation.

6. RandomAlgorithm Class: A simpler version that just tries random combinations of items to find the best one.


In the end, the code helps to find the best selection of items to place on a shelf that maximizes profit without going over the shelf space limit.

In [36]:
# Code 1: Retail Store Items
retail_store_items = {
    "Toothpaste": {"shelf_space": 1, "expected_profit": 2.50},
    "Shampoo": {"shelf_space": 3, "expected_profit": 3.75},
    "Conditioner": {"shelf_space": 3, "expected_profit": 3.50},
    "Body Wash": {"shelf_space": 3, "expected_profit": 3.00},
    "Deodorant": {"shelf_space": 1, "expected_profit": 2.20},
    "Lotion": {"shelf_space": 3, "expected_profit": 3.00},
    "Shaving Cream": {"shelf_space": 2, "expected_profit": 2.50},
    "Razor Blades": {"shelf_space": 1, "expected_profit": 4.50},
    "Mouthwash": {"shelf_space": 2, "expected_profit": 2.00},
    "Hand Soap": {"shelf_space": 1, "expected_profit": 1.50},
    "Facial Tissue": {"shelf_space": 2, "expected_profit": 1.20},
    "Paper Towels": {"shelf_space": 5, "expected_profit": 3.00},
    "Toilet Paper": {"shelf_space": 5, "expected_profit": 2.80},
    "Dish Soap": {"shelf_space": 3, "expected_profit": 2.75},
    "Laundry Detergent": {"shelf_space": 6, "expected_profit": 4.50},
    "Fabric Softener": {"shelf_space": 3, "expected_profit": 3.50},
    "Bleach": {"shelf_space": 4, "expected_profit": 2.00},
    "All-Purpose Cleaner": {"shelf_space": 3, "expected_profit": 2.50},
    "Glass Cleaner": {"shelf_space": 2, "expected_profit": 2.00},
    "Furniture Polish": {"shelf_space": 2, "expected_profit": 2.50},
    "Trash Bags": {"shelf_space": 4, "expected_profit": 3.00},
    "Light Bulbs": {"shelf_space": 2, "expected_profit": 1.50},
    "Batteries": {"shelf_space": 1, "expected_profit": 2.00},
    "Extension Cords": {"shelf_space": 3, "expected_profit": 4.00},
    "Pet Food": {"shelf_space": 6, "expected_profit": 3.50},
    "Cat Litter": {"shelf_space": 5, "expected_profit": 4.00},
    "Dog Leash": {"shelf_space": 2, "expected_profit": 3.00},
    "Pet Shampoo": {"shelf_space": 2, "expected_profit": 2.50},
    "Pet Toys": {"shelf_space": 2, "expected_profit": 1.50},
    "Shower Curtain": {"shelf_space": 4, "expected_profit": 3.00},
    "Bath Towels": {"shelf_space": 4, "expected_profit": 4.00},
    "Bed Sheets": {"shelf_space": 6, "expected_profit": 8.00},
    "Pillows": {"shelf_space": 5, "expected_profit": 5.00},
    "Blankets": {"shelf_space": 6, "expected_profit": 6.00},
    "Curtains": {"shelf_space": 5, "expected_profit": 7.00},
    "Wall Clocks": {"shelf_space": 3, "expected_profit": 4.50},
    "Picture Frames": {"shelf_space": 2, "expected_profit": 2.00},
    "Candles": {"shelf_space": 2, "expected_profit": 2.50},
    "Vases": {"shelf_space": 3, "expected_profit": 3.00},
    "Decorative Pillows": {"shelf_space": 4, "expected_profit": 3.50},
    "Alarm Clocks": {"shelf_space": 2, "expected_profit": 2.00},
    "Phone Chargers": {"shelf_space": 1, "expected_profit": 3.00},
    "USB Cables": {"shelf_space": 1, "expected_profit": 1.50},
    "Headphones": {"shelf_space": 2, "expected_profit": 5.00},
    "Laptop Sleeves": {"shelf_space": 2, "expected_profit": 4.00},
    "Notebooks": {"shelf_space": 2, "expected_profit": 1.50},
    "Pens": {"shelf_space": 1, "expected_profit": 1.00},
    "Sticky Notes": {"shelf_space": 1, "expected_profit": 1.00},
    "Folders": {"shelf_space": 2, "expected_profit": 1.50},
    "Scissors": {"shelf_space": 1, "expected_profit": 2.00}
}

# Code 2: Import Required Libraries
import random
from tqdm import tqdm

# Code 3: Individual Class
class Individual:
    def __init__(self, merch, max_shelf_space):
        self.merch = merch
        self.max_shelf_space = max_shelf_space
        self.chromosome = [random.choice([0, 1]) for _ in range(len(merch))]
        self.occupied_space = 0
        self.total_expected_profit = 0
        self.generation = 0
        self.fitness()  # Calculate fitness on initialization

    def fitness(self):
        self.occupied_space = 0
        self.total_expected_profit = 0
        for (item, values), multiplier in zip(self.merch.items(), self.chromosome):
            self.occupied_space += values['shelf_space'] * multiplier
            self.total_expected_profit += values['expected_profit'] * multiplier
            if self.occupied_space > self.max_shelf_space:
                self.total_expected_profit = 0
                break

    def crossover(self, partner):
        cutoff_1 = random.randint(1, len(self.chromosome) - 1)
        cutoff_2 = random.randint(1, len(self.chromosome) - 1)
        if cutoff_1 > cutoff_2:
            cutoff_1, cutoff_2 = cutoff_2, cutoff_1

        child1 = Individual(self.merch, self.max_shelf_space)
        child2 = Individual(self.merch, self.max_shelf_space)
        child1.chromosome = self.chromosome[:cutoff_1] + partner.chromosome[cutoff_1:cutoff_2] + self.chromosome[cutoff_2:]
        child2.chromosome = partner.chromosome[:cutoff_1] + self.chromosome[cutoff_1:cutoff_2] + partner.chromosome[cutoff_2:]

        child1.fitness()
        child2.fitness()

        child1.generation = self.generation + 1
        child2.generation = self.generation + 1

        return child1, child2

    def mutation(self, rate):
        if random.random() < rate:
            random_position = random.randint(0, len(self.chromosome) - 1)
            self.chromosome[random_position] = 1 - self.chromosome[random_position]
            self.fitness()

    def view_shelf(self):
        for (item, values), multiplier in zip(self.merch.items(), self.chromosome):
            if multiplier == 1:
                print(item)

# Code 8: GeneticAlgorithm Class
class GeneticAlgorithm:
    def __init__(self, population_size, num_of_generations, mutation_rate, max_shelf_space, items):
        self.population_size = population_size
        self.num_of_generations = num_of_generations
        self.mutation_rate = mutation_rate
        self.max_shelf_space = max_shelf_space
        self.items = items
        self.best_expected_profit = 0
        self.best_individuals = []
        self.population = []

    def run_simulation(self):
        #### Initial Population ####
        for i in range(self.population_size):
            individual = Individual(merch=self.items, max_shelf_space=self.max_shelf_space)
            self.population.append(individual)
            if individual.total_expected_profit > self.best_expected_profit:
                self.best_expected_profit = individual.total_expected_profit
                self.best_individuals.append(individual)

        #### Subsequent Populations ####
        for g in tqdm(range(self.num_of_generations)):
            random.shuffle(self.population)
            new_population = []

            #### Crossover between each pair ####
            for pair in range(0, len(self.population), 2):
                parent_1 = self.population[pair]
                parent_2 = self.population[pair + 1]
                child_1, child_2 = parent_1.crossover(parent_2)
                child_1.mutation(rate=self.mutation_rate)
                child_2.mutation(rate=self.mutation_rate)
                all_children = [child_1, child_2]
                all_children.sort(key=lambda x: x.total_expected_profit, reverse=True)
                top_children = all_children[:2]
                if top_children[0].total_expected_profit > self.best_expected_profit:
                    self.best_expected_profit = top_children[0].total_expected_profit
                    self.best_individuals.append(top_children[0])
                if top_children[1].total_expected_profit > self.best_expected_profit:
                    self.best_expected_profit = top_children[1].total_expected_profit
                    self.best_individuals.append(top_children[1])
                new_population.append(top_children[0])
                new_population.append(top_children[1])

            #### Assign population as the new population ####
            self.population = new_population

# Code 10: RandomAlgorithm Class
class RandomAlgorithm:
    def __init__(self, iterations, max_shelf_space, items):
        self.iterations = iterations
        self.max_shelf_space = max_shelf_space
        self.items = items
        self.best_individuals = []
        self.best_expected_profit = 0

    def run_simulation(self):
        #### Initial Population ####
        for i in tqdm(range(self.iterations)):
            trial = Individual(merch=self.items, max_shelf_space=self.max_shelf_space)
            trial.fitness()
            if trial.total_expected_profit > self.best_expected_profit:
                self.best_expected_profit = trial.total_expected_profit
                self.best_individuals.append(trial)



**Trial Run 1 — 1,000 instances**


Now, let's do our first trial run. We will run our Genetic Algorithm through 10 generations, with each population being of size 100, which yields 1,000 total Individual classes created.

In addition, we will designate our Random Algorithm to generate 1,000 classes. Note that our max_shelf_space will be 100. Check out the results below!

In [37]:
# Code 11: Run Random Algorithm
ra = RandomAlgorithm(iterations=1000, max_shelf_space=100, items=retail_store_items)
ra.run_simulation()



100%|██████████| 1000/1000 [00:00<00:00, 4899.06it/s]


In [38]:
ra.best_expected_profit

107.2

In [39]:
# Code 12: Run Genetic Algorithm
GA = GeneticAlgorithm(population_size=100, num_of_generations=10, mutation_rate=0.01, items=retail_store_items, max_shelf_space=100)
GA.run_simulation()

100%|██████████| 10/10 [00:00<00:00, 27.20it/s]


In [40]:
GA.best_expected_profit

110.7

**Our Genetic Algorithm won the first round** with the slight best-expected profit of 107 vs. 107.25! Let's see if this gap increases as we increase the iterations.

Next up, let's try out 10,000 instances.

**Trial Run 2— 10,000 instances**

In [41]:
ra = RandomAlgorithm(iterations=10000, max_shelf_space=100, items=retail_store_items)
ra.run_simulation()

100%|██████████| 10000/10000 [00:02<00:00, 4820.83it/s]


In [42]:
ra.best_expected_profit

112.45

In [43]:
GA = GeneticAlgorithm(population_size=100, num_of_generations=100, mutation_rate=0.01, items=retail_store_items, max_shelf_space=100)
GA.run_simulation()

100%|██████████| 100/100 [00:01<00:00, 74.20it/s]


In [44]:
GA.best_expected_profit

111.4

Once again, our Genetic Algorithm won! I am curious to see how this result varies if we use a population size of 1,000 with 10 generations instead of 100 and 100 generations. Let's see what happens.

In [45]:
GA = GeneticAlgorithm(population_size=1000, num_of_generations=10, mutation_rate=0.01, items=retail_store_items, max_shelf_space=100)
GA.run_simulation()

100%|██████████| 10/10 [00:00<00:00, 12.30it/s]


In [46]:
GA.best_expected_profit

111.0

Our expected profit went down. Would this always be the case? I have some ideas on how to test this, but we will explore that in a future article! For now, let's keep the focus on testing the Genetic Algorithm vs. the random approach.

Finally, let's see what happens with 100,000 iterations. Our Genetic Algorithm will have populations of 1,000 and 100 generations.

**Trial Run 3–100,000 instances**

In [47]:
ra = RandomAlgorithm(iterations=100000, max_shelf_space=100, items=retail_store_items)
ra.run_simulation()

100%|██████████| 100000/100000 [00:07<00:00, 12850.92it/s]


In [48]:
ra.best_expected_profit

115.8

In [49]:
GA = GeneticAlgorithm(population_size=1000, num_of_generations=100, mutation_rate=0.01, items=retail_store_items, max_shelf_space=100)
GA.run_simulation()

100%|██████████| 100/100 [00:09<00:00, 10.04it/s]


In [50]:
GA.best_expected_profit

113.4

Although our Genetic Algorithm won, the expected profit seems to oscillate between 117 and 125. Luckily, our Genetic Algorithm object saves each instance of the Individual who becomes the new "best Individual." Using this information, we can visualize the generations where a new "best Individual" was revealed and their expected profits. This will help us see where our algorithm converges. I've listed out the code below to plot this using Seaborn.